mini/concurrency_test.go

```diff
@@ -0,0 +1,208 @@
+package mini
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/awalterschulze/gominikanren/micro"
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+// remember, concurrency != parallelism
+
+// setting up exclusive-or for a very simple or-parallelism demo
+func disjExcl(g1, g2 micro.Goal) micro.Goal {
+	return func() micro.GoalFn {
+		return func(s *micro.State) micro.StreamOfStates {
+			g1s := g1()(s)
+			if g1s != nil {
+				return g1s
+			}
+			g2s := g2()(s)
+			return micro.Mplus(g1s, g2s)
+		}
+	}
+}
+
+func TestDisjExcl(t *testing.T) {
+	sexprs := micro.Run(-1, func(q *ast.SExpr) micro.Goal {
+		return disjExcl(
+			micro.EqualO(q, ast.NewInt(1)),
+			micro.EqualO(q, ast.NewInt(2)),
+		)
+	})
+	if len(sexprs) != 1 {
+		t.Fatalf("expected len %d, but got len %d instead", 1, len(sexprs))
+	}
+	for _, sexpr := range sexprs {
+		if sexpr.String() != "1" {
+			t.Fatalf("expected %s, but got %s instead", "1", sexpr.String())
+		}
+	}
+}
+
+// note: this is a naive implementation with runaway goroutines;
+// in a proper implementation the goroutines are taken from a pool
+// and properly cancelled when returning early
+func concurrentDisjExcl(g1, g2 micro.Goal) micro.Goal {
+	return func() micro.GoalFn {
+		return func(s *micro.State) micro.StreamOfStates {
+			c1 := make(chan micro.StreamOfStates, 1)
+			c2 := make(chan micro.StreamOfStates, 1)
+			go func() {
+				c1 <- g1()(s)
+			}()
+			go func() {
+				c2 <- g2()(s)
+			}()
+			var g1s, g2s micro.StreamOfStates
+			for i := 0; i < 2; i++ {
```

Comment:
deriveJoin can do this, if you also call `close` inside the go routines.
See here for example https://github.com/awalterschulze/goderive/blob/master/test/normal/join_test.go#L235